ICE: `$ident: found TraitItem(TraitItem `

<!--
[31mICE[0m: Rustc ./a.rs '--edition=2024 -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' panicked at compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs:210:61: '$ident: found TraitItem(TraitItem { ident: A#0, owner_id: DefId(0:4 ~ a[80eb]::TraitWAssocConst::A), generics: Generics { params: [], predicates: [], has_where_clause_predicates: false, where_clause_span: /tmp/im/a.rs:3:33: 3:33 (#0), span: /tmp/im/a.rs:3:10: 3:10 (#0) }, kind: Const(Ty { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).8), span: /tmp/im/a.rs:3:12: 3:33 (#0), kind: TraitObject([PolyTraitRef { bound_generic_params: [], modifiers: TraitBoundModifiers { constness: Never, polarity: Positive }, trait_ref: TraitRef { path: Path { span: /tmp/im/a.rs:3:12: 3:33 (#0), res: Def(Trait, DefId(0:3 ~ a[80eb]::TraitWAssocConst)), segments: [PathSegment { ident: TraitWAssocConst#0, hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).5), res: Def(Trait, DefId(0:3 ~ a[80eb]::TraitWAssocConst)), args: Some(GenericArgs { args: [], constraints: [AssocItemConstraint { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).4), ident: A#0, gen_args: GenericArgs { args: [], constraints: [], parenthesized: No, span_ext: no-location (#0) }, kind: Equality { term: Const(ConstArg { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).3), kind: Anon(AnonConst { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).1), def_id: DefId(0:5 ~ a[80eb]::TraitWAssocConst::A::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).2) }, span: /tmp/im/a.rs:3:31: 3:32 (#0) }) }) }, span: /tmp/im/a.rs:3:29: 3:32 (#0) }], parenthesized: No, span_ext: /tmp/im/a.rs:3:28: 3:33 (#0) }), infer_args: false }] }, hir_ref_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).6) }, span: /tmp/im/a.rs:3:12: 3:33 (#0) }], TaggedRef { pointer: Lifetime { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).7), ident: #0, res: ImplicitObjectLifetimeDefault }, tag: None }) }, None), span: /tmp/im/a.rs:3:3: 3:34 (#0), defaultness: Default { has_value: false } })'', 'thread 'rustc' panicked at compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs:210:61: '$ident: found TraitItem(TraitItem { ident: A#0, owner_id: DefId(0:4 ~ a[80eb]::TraitWAssocConst::A), generics: Generics { params: [], predicates: [], has_where_clause_predicates: false, where_clause_span: /tmp/im/a.rs:3:33: 3:33 (#0), span: /tmp/im/a.rs:3:10: 3:10 (#0) }, kind: Const(Ty { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).8), span: /tmp/im/a.rs:3:12: 3:33 (#0), kind: TraitObject([PolyTraitRef { bound_generic_params: [], modifiers: TraitBoundModifiers { constness: Never, polarity: Positive }, trait_ref: TraitRef { path: Path { span: /tmp/im/a.rs:3:12: 3:33 (#0), res: Def(Trait, DefId(0:3 ~ a[80eb]::TraitWAssocConst)), segments: [PathSegment { ident: TraitWAssocConst#0, hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).5), res: Def(Trait, DefId(0:3 ~ a[80eb]::TraitWAssocConst)), args: Some(GenericArgs { args: [], constraints: [AssocItemConstraint { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).4), ident: A#0, gen_args: GenericArgs { args: [], constraints: [], parenthesized: No, span_ext: no-location (#0) }, kind: Equality { term: Const(ConstArg { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).3), kind: Anon(AnonConst { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).1), def_id: DefId(0:5 ~ a[80eb]::TraitWAssocConst::A::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).2) }, span: /tmp/im/a.rs:3:31: 3:32 (#0) }) }) }, span: /tmp/im/a.rs:3:29: 3:32 (#0) }], parenthesized: No, span_ext: /tmp/im/a.rs:3:28: 3:33 (#0) }), infer_args: false }] }, hir_ref_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).6) }, span: /tmp/im/a.rs:3:12: 3:33 (#0) }], TaggedRef { pointer: Lifetime { hir_id: HirId(DefId(0:4 ~ a[80eb]::TraitWAssocConst::A).7), ident: #0, res: ImplicitObjectLifetimeDefault }, tag: None }) }, None), span: /tmp/im/a.rs:3:3: 3:34 (#0), defaultness: Default { has_value: false } })''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: --edition=2024
pub trait TraitWAssocConst {
    const A: TraitWAssocConst<A = 0>;
}

````

original:
````rust
#![feature(generic_const_exprs)]
pub trait TraitWAssocConst {
  const A: TraitWAssocConst<A=0>;
}

fn main<A: TraitWAssocConst<A=32>>() {
  TraitWAssocConst<A=0>:: TraitWAssocConst<A=32> ();
}

````

Version information
````
rustc 1.87.0-nightly (20f0108ad 2025-03-08)
binary: rustc
commit-hash: 20f0108ada90e37c2350ad68fdc6afe56f51539d
commit-date: 2025-03-08
host: x86_64-unknown-linux-gnu
release: 1.87.0-nightly
LLVM version: 20.1.0

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/20f0108ada90e37c2350ad68fdc6afe56f51539d/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs#L204-L216

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc --edition=2024`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0658]: associated const equality is incomplete
 --> /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:31
  |
2 |     const A: TraitWAssocConst<A = 0>;
  |                               ^^^^^
  |
  = note: see issue #92827 <https://github.com/rust-lang/rust/issues/92827> for more information
  = help: add `#![feature(associated_const_equality)]` to the crate attributes to enable
  = note: this compiler was built on 2025-03-08; consider upgrading it if it is out of date

error[E0601]: `main` function not found in crate `mvce`
 --> /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:3:2
  |
3 | }
  |  ^ consider adding a `main` function to `/tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs`


thread 'rustc' panicked at compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs:210:61:
$ident: found TraitItem(TraitItem { ident: A#0, owner_id: DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A), generics: Generics { params: [], predicates: [], has_where_clause_predicates: false, where_clause_span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:37: 2:37 (#0), span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:12: 2:12 (#0) }, kind: Const(Ty { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).8), span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:14: 2:37 (#0), kind: TraitObject([PolyTraitRef { bound_generic_params: [], modifiers: TraitBoundModifiers { constness: Never, polarity: Positive }, trait_ref: TraitRef { path: Path { span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:14: 2:37 (#0), res: Def(Trait, DefId(0:3 ~ mvce[a2ab]::TraitWAssocConst)), segments: [PathSegment { ident: TraitWAssocConst#0, hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).5), res: Def(Trait, DefId(0:3 ~ mvce[a2ab]::TraitWAssocConst)), args: Some(GenericArgs { args: [], constraints: [AssocItemConstraint { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).4), ident: A#0, gen_args: GenericArgs { args: [], constraints: [], parenthesized: No, span_ext: no-location (#0) }, kind: Equality { term: Const(ConstArg { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).3), kind: Anon(AnonConst { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).1), def_id: DefId(0:5 ~ mvce[a2ab]::TraitWAssocConst::A::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).2) }, span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:35: 2:36 (#0) }) }) }, span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:31: 2:36 (#0) }], parenthesized: No, span_ext: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:30: 2:37 (#0) }), infer_args: false }] }, hir_ref_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).6) }, span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:14: 2:37 (#0) }], TaggedRef { pointer: Lifetime { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).7), ident: #0, res: ImplicitObjectLifetimeDefault }, tag: None }) }, None), span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:5: 2:38 (#0), defaultness: Default { has_value: false } })
stack backtrace:
   0:     0x735d0f5c6f34 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h07b2089c80742ed5
   1:     0x735d0fe05aa2 - core::fmt::write::h332ae571ae4a9cc4
   2:     0x735d111e3a11 - std::io::Write::write_fmt::h0c761025205b813c
   3:     0x735d0f5c6d92 - std::sys::backtrace::BacktraceLock::print::hbba93a9383b9fa8a
   4:     0x735d0f5c9672 - std::panicking::default_hook::{{closure}}::h00e63950a63732af
   5:     0x735d0f5c9264 - std::panicking::default_hook::h400dc9bc1b724c37
   6:     0x735d0e7204e7 - std[c20316ad899fcaf5]::panicking::update_hook::<alloc[2db8fd01bd7ea837]::boxed::Box<rustc_driver_impl[153d3eccf0c7c750]::install_ice_hook::{closure#1}>>::{closure#0}
   7:     0x735d0f5c9ee3 - std::panicking::rust_panic_with_hook::he428532615a29ae4
   8:     0x735d0f5c9bda - std::panicking::begin_panic_handler::{{closure}}::h498acb37174be106
   9:     0x735d0f5c73f9 - std::sys::backtrace::__rust_end_short_backtrace::h2c0df5d26e60f09c
  10:     0x735d0f5c989d - rust_begin_unwind
  11:     0x735d0c25df10 - core::panicking::panic_fmt::h35b514d9ffb031f4
  12:     0x735d0e7ecf81 - rustc_hir[229de859651313c0]::hir::expect_failed::<&rustc_hir[229de859651313c0]::hir::Node>
  13:     0x735d1105f858 - <dyn rustc_hir_analysis[2ac45909935d36a1]::hir_ty_lowering::HirTyLowerer>::prohibit_or_lint_bare_trait_object_ty
  14:     0x735d10566736 - <dyn rustc_hir_analysis[2ac45909935d36a1]::hir_ty_lowering::HirTyLowerer>::lower_ty
  15:     0x735d103618cd - rustc_hir_analysis[2ac45909935d36a1]::collect::type_of::type_of
  16:     0x735d10359d2a - rustc_query_impl[824584145395502d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[824584145395502d]::query_impl::type_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 8usize]>>
  17:     0x735d10161332 - rustc_query_system[75007086aa92bf3]::query::plumbing::try_execute_query::<rustc_query_impl[824584145395502d]::DynamicConfig<rustc_query_system[75007086aa92bf3]::query::caches::DefIdCache<rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[824584145395502d]::plumbing::QueryCtxt, false>
  18:     0x735d10160edf - rustc_query_impl[824584145395502d]::query_impl::type_of::get_query_non_incr::__rust_end_short_backtrace
  19:     0x735d10597699 - <rustc_hir_analysis[2ac45909935d36a1]::collect::CollectItemTypesVisitor as rustc_hir[229de859651313c0]::intravisit::Visitor>::visit_trait_item
  20:     0x735d105899ef - rustc_hir_analysis[2ac45909935d36a1]::check::wfcheck::check_well_formed
  21:     0x735d1058754d - rustc_query_impl[824584145395502d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[824584145395502d]::query_impl::check_well_formed::dynamic_query::{closure#2}::{closure#0}, rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 1usize]>>
  22:     0x735d10587004 - rustc_query_system[75007086aa92bf3]::query::plumbing::try_execute_query::<rustc_query_impl[824584145395502d]::DynamicConfig<rustc_data_structures[59c1b7daa758d1f8]::vec_cache::VecCache<rustc_span[268b27c29cacb432]::def_id::LocalDefId, rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 1usize]>, rustc_query_system[75007086aa92bf3]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[824584145395502d]::plumbing::QueryCtxt, false>
  23:     0x735d10586abf - rustc_query_impl[824584145395502d]::query_impl::check_well_formed::get_query_non_incr::__rust_end_short_backtrace
  24:     0x735d105845f0 - rustc_hir_analysis[2ac45909935d36a1]::check::wfcheck::check_mod_type_wf
  25:     0x735d105840df - rustc_query_impl[824584145395502d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[824584145395502d]::query_impl::check_mod_type_wf::dynamic_query::{closure#2}::{closure#0}, rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 1usize]>>
  26:     0x735d10e96d56 - rustc_query_system[75007086aa92bf3]::query::plumbing::try_execute_query::<rustc_query_impl[824584145395502d]::DynamicConfig<rustc_query_system[75007086aa92bf3]::query::caches::DefaultCache<rustc_span[268b27c29cacb432]::def_id::LocalModDefId, rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[824584145395502d]::plumbing::QueryCtxt, false>
  27:     0x735d10e96aff - rustc_query_impl[824584145395502d]::query_impl::check_mod_type_wf::get_query_non_incr::__rust_end_short_backtrace
  28:     0x735d1017f9cf - rustc_hir_analysis[2ac45909935d36a1]::check_crate
  29:     0x735d1032807a - rustc_interface[7e29b5e2b62c504d]::passes::run_required_analyses
  30:     0x735d10dc1b7a - rustc_interface[7e29b5e2b62c504d]::passes::analysis
  31:     0x735d10dc1b59 - rustc_query_impl[824584145395502d]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[824584145395502d]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 0usize]>>
  32:     0x735d10eda3cb - rustc_query_system[75007086aa92bf3]::query::plumbing::try_execute_query::<rustc_query_impl[824584145395502d]::DynamicConfig<rustc_query_system[75007086aa92bf3]::query::caches::SingleCache<rustc_middle[b1dca1d1b343b3c]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[824584145395502d]::plumbing::QueryCtxt, false>
  33:     0x735d10eda0b9 - rustc_query_impl[824584145395502d]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  34:     0x735d10f1577d - rustc_interface[7e29b5e2b62c504d]::passes::create_and_enter_global_ctxt::<core[8e02e1de5014891c]::option::Option<rustc_interface[7e29b5e2b62c504d]::queries::Linker>, rustc_driver_impl[153d3eccf0c7c750]::run_compiler::{closure#0}::{closure#2}>::{closure#2}::{closure#0}
  35:     0x735d10f056a0 - rustc_interface[7e29b5e2b62c504d]::interface::run_compiler::<(), rustc_driver_impl[153d3eccf0c7c750]::run_compiler::{closure#0}>::{closure#1}
  36:     0x735d10dabe48 - std[c20316ad899fcaf5]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[7e29b5e2b62c504d]::util::run_in_thread_with_globals<rustc_interface[7e29b5e2b62c504d]::util::run_in_thread_pool_with_globals<rustc_interface[7e29b5e2b62c504d]::interface::run_compiler<(), rustc_driver_impl[153d3eccf0c7c750]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  37:     0x735d10dac734 - <<std[c20316ad899fcaf5]::thread::Builder>::spawn_unchecked_<rustc_interface[7e29b5e2b62c504d]::util::run_in_thread_with_globals<rustc_interface[7e29b5e2b62c504d]::util::run_in_thread_pool_with_globals<rustc_interface[7e29b5e2b62c504d]::interface::run_compiler<(), rustc_driver_impl[153d3eccf0c7c750]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[8e02e1de5014891c]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  38:     0x735d10dadb2b - std::sys::pal::unix::thread::Thread::new::thread_start::h63137a35d539f669
  39:     0x735d0aea370a - <unknown>
  40:     0x735d0af27aac - <unknown>
  41:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.87.0-nightly (20f0108ad 2025-03-08) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z dump-mir-dir=dir

query stack during panic:
#0 [type_of] computing type of `TraitWAssocConst::A`
#1 [check_well_formed] checking that `TraitWAssocConst::A` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0601, E0658.
For more information about an error, try `rustc --explain E0601`.

```

</p>
</details>

<!--
query stack:
$ident: found TraitItem(TraitItem { ident: A#0, owner_id: DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A), generics: Generics { params: [], predicates: [], has_where_clause_predicates: false, where_clause_span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:37: 2:37 (#0), span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:12: 2:12 (#0) }, kind: Const(Ty { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).8), span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:14: 2:37 (#0), kind: TraitObject([PolyTraitRef { bound_generic_params: [], modifiers: TraitBoundModifiers { constness: Never, polarity: Positive }, trait_ref: TraitRef { path: Path { span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:14: 2:37 (#0), res: Def(Trait, DefId(0:3 ~ mvce[a2ab]::TraitWAssocConst)), segments: [PathSegment { ident: TraitWAssocConst#0, hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).5), res: Def(Trait, DefId(0:3 ~ mvce[a2ab]::TraitWAssocConst)), args: Some(GenericArgs { args: [], constraints: [AssocItemConstraint { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).4), ident: A#0, gen_args: GenericArgs { args: [], constraints: [], parenthesized: No, span_ext: no-location (#0) }, kind: Equality { term: Const(ConstArg { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).3), kind: Anon(AnonConst { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).1), def_id: DefId(0:5 ~ mvce[a2ab]::TraitWAssocConst::A::{constant#0}), body: BodyId { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).2) }, span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:35: 2:36 (#0) }) }) }, span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:31: 2:36 (#0) }], parenthesized: No, span_ext: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:30: 2:37 (#0) }), infer_args: false }] }, hir_ref_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).6) }, span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:14: 2:37 (#0) }], TaggedRef { pointer: Lifetime { hir_id: HirId(DefId(0:4 ~ mvce[a2ab]::TraitWAssocConst::A).7), ident: #0, res: ImplicitObjectLifetimeDefault }, tag: None }) }, None), span: /tmp/icemaker_global_tempdir.xCmlSFx4hmpo/rustc_testrunner_tmpdir_reporting.tPdpsshuU7cG/mvce.rs:2:5: 2:38 (#0), defaultness: Default { has_value: false } })
#0 [type_of] computing type of `TraitWAssocConst::A`
#1 [check_well_formed] checking that `TraitWAssocConst::A` is well-formed
#2 [check_mod_type_wf] checking that types are well-formed in top-level module
#3 [analysis] running analysis passes on this crate

-->

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs
@@ -123,6 +123,8 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         }
     }
 
+    /// For a struct or enum with an invalid bare trait object field, suggest turning
+    /// it into a generic type bound.
     fn maybe_suggest_add_generic_impl_trait(
         &self,
         self_ty: &hir::Ty<'_>,
@@ -132,21 +134,38 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         let msg = "you might be missing a type parameter";
         let mut sugg = vec![];
 
-        let parent_id = tcx.hir_get_parent_item(self_ty.hir_id).def_id;
-        let parent_item = tcx.hir_node_by_def_id(parent_id).expect_item();
-        match parent_item.kind {
-            hir::ItemKind::Struct(_, generics) | hir::ItemKind::Enum(_, generics) => {
-                sugg.push((
-                    generics.where_clause_span,
-                    format!(
-                        "<T: {}>",
-                        self.tcx().sess.source_map().span_to_snippet(self_ty.span).unwrap()
-                    ),
-                ));
-                sugg.push((self_ty.span, "T".to_string()));
+        let parent_hir_id = tcx.parent_hir_id(self_ty.hir_id);
+        let parent_item = tcx.hir_get_parent_item(self_ty.hir_id).def_id;
+
+        let generics = match tcx.hir_node_by_def_id(parent_item) {
+            hir::Node::Item(hir::Item {
+                kind: hir::ItemKind::Struct(variant, generics), ..
+            }) => {
+                if !variant.fields().iter().any(|field| field.hir_id == parent_hir_id) {
+                    return false;
+                }
+                generics
             }
-            _ => {}
-        }
+            hir::Node::Item(hir::Item { kind: hir::ItemKind::Enum(def, generics), .. }) => {
+                if !def
+                    .variants
+                    .iter()
+                    .flat_map(|variant| variant.data.fields().iter())
+                    .any(|field| field.hir_id == parent_hir_id)
+                {
+                    return false;
+                }
+                generics
+            }
+            _ => return false,
+        };
+
+        // FIXME: `T` may already be taken.
+        sugg.push((
+            generics.where_clause_span,
+            format!("<T: {}>", self.tcx().sess.source_map().span_to_snippet(self_ty.span).unwrap()),
+        ));
+        sugg.push((self_ty.span, "T".to_string()));
         diag.multipart_suggestion_verbose(msg, sugg, Applicability::MachineApplicable);
         true
     }
@@ -198,6 +217,13 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         }
     }
 
+    /// Try our best to approximate when adding `dyn` would be helpful for a bare
+    /// trait object.
+    ///
+    /// Right now, this is if the type is either directly nested in another ty,
+    /// or if it's in the tail field within a struct. This approximates what the
+    /// user would've gotten on edition 2015, except for the case where we have
+    /// an *obvious* knock-on `Sized` error.
     fn maybe_suggest_dyn_trait(
         &self,
         self_ty: &hir::Ty<'_>,
@@ -206,19 +232,40 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         diag: &mut Diag<'_>,
     ) -> bool {
         let tcx = self.tcx();
-        let parent_id = tcx.hir_get_parent_item(self_ty.hir_id).def_id;
-        let parent_item = tcx.hir_node_by_def_id(parent_id).expect_item();
 
-        // If the parent item is an enum, don't suggest the dyn trait.
-        if let hir::ItemKind::Enum(..) = parent_item.kind {
-            return false;
-        }
+        // Look at the direct HIR parent, since we care about the relationship between
+        // the type and the thing that directly encloses it.
+        match tcx.parent_hir_node(self_ty.hir_id) {
+            // These are all generally ok. Namely, when a trait object is nested
+            // into another expression or ty, it's either very certain that they
+            // missed the ty (e.g. `&Trait`) or it's not really possible to tell
+            // what their intention is, so let's not give confusing suggestions and
+            // just mention `dyn`. The user can make up their mind what to do here.
+            hir::Node::Ty(_)
+            | hir::Node::Expr(_)
+            | hir::Node::PatExpr(_)
+            | hir::Node::PathSegment(_)
+            | hir::Node::AssocItemConstraint(_)
+            | hir::Node::TraitRef(_)
+            | hir::Node::Item(_)
+            | hir::Node::WherePredicate(_) => {}
 
-        // If the parent item is a struct, check if self_ty is the last field.
-        if let hir::ItemKind::Struct(variant_data, _) = parent_item.kind {
-            if variant_data.fields().last().unwrap().ty.span != self_ty.span {
-                return false;
+            hir::Node::Field(field) => {
+                // Enums can't have unsized fields, fields can only have an unsized tail field.
+                if let hir::Node::Item(hir::Item {
+                    kind: hir::ItemKind::Struct(variant, _), ..
+                }) = tcx.parent_hir_node(field.hir_id)
+                    && variant
+                        .fields()
+                        .last()
+                        .is_some_and(|tail_field| tail_field.hir_id == field.hir_id)
+                {
+                    // Ok
+                } else {
+                    return false;
+                }
             }
+            _ => return false,
         }
 
         // FIXME: Only emit this suggestion if the trait is dyn-compatible.

--- a/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs
+++ b/compiler/rustc_hir_analysis/src/hir_ty_lowering/lint.rs
@@ -78,13 +78,16 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         }
 
         if self_ty.span.edition().at_least_rust_2021() {
-            let msg = "expected a type, found a trait";
-            let label = "you can add the `dyn` keyword if you want a trait object";
-            let mut diag =
-                rustc_errors::struct_span_code_err!(self.dcx(), self_ty.span, E0782, "{}", msg);
+            let mut diag = rustc_errors::struct_span_code_err!(
+                self.dcx(),
+                self_ty.span,
+                E0782,
+                "{}",
+                "expected a type, found a trait"
+            );
             if self_ty.span.can_be_used_for_suggestions()
                 && !self.maybe_suggest_impl_trait(self_ty, &mut diag)
-                && !self.maybe_suggest_dyn_trait(self_ty, label, sugg, &mut diag)
+                && !self.maybe_suggest_dyn_trait(self_ty, sugg, &mut diag)
             {
                 self.maybe_suggest_add_generic_impl_trait(self_ty, &mut diag);
             }
@@ -131,8 +134,6 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         diag: &mut Diag<'_>,
     ) -> bool {
         let tcx = self.tcx();
-        let msg = "you might be missing a type parameter";
-        let mut sugg = vec![];
 
         let parent_hir_id = tcx.parent_hir_id(self_ty.hir_id);
         let parent_item = tcx.hir_get_parent_item(self_ty.hir_id).def_id;
@@ -160,13 +161,27 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
             _ => return false,
         };
 
-        // FIXME: `T` may already be taken.
-        sugg.push((
-            generics.where_clause_span,
-            format!("<T: {}>", self.tcx().sess.source_map().span_to_snippet(self_ty.span).unwrap()),
-        ));
-        sugg.push((self_ty.span, "T".to_string()));
-        diag.multipart_suggestion_verbose(msg, sugg, Applicability::MachineApplicable);
+        let Ok(rendered_ty) = tcx.sess.source_map().span_to_snippet(self_ty.span) else {
+            return false;
+        };
+
+        let param = "TUV"
+            .chars()
+            .map(|c| c.to_string())
+            .chain((0..).map(|i| format!("P{i}")))
+            .find(|s| !generics.params.iter().any(|param| param.name.ident().as_str() == s))
+            .expect("we definitely can find at least one param name to generate");
+        let mut sugg = vec![(self_ty.span, param.to_string())];
+        if let Some(insertion_span) = generics.span_for_param_suggestion() {
+            sugg.push((insertion_span, format!(", {param}: {}", rendered_ty)));
+        } else {
+            sugg.push((generics.where_clause_span, format!("<{param}: {}>", rendered_ty)));
+        }
+        diag.multipart_suggestion_verbose(
+            "you might be missing a type parameter",
+            sugg,
+            Applicability::MachineApplicable,
+        );
         true
     }
     /// Make sure that we are in the condition to suggest the blanket implementation.
@@ -227,7 +242,6 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     fn maybe_suggest_dyn_trait(
         &self,
         self_ty: &hir::Ty<'_>,
-        label: &str,
         sugg: Vec<(Span, String)>,
         diag: &mut Diag<'_>,
     ) -> bool {
@@ -270,7 +284,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
 
         // FIXME: Only emit this suggestion if the trait is dyn-compatible.
         diag.multipart_suggestion_verbose(
-            label.to_string(),
+            "you can add the `dyn` keyword if you want a trait object",
             sugg,
             Applicability::MachineApplicable,
         );

--- a/tests/ui/dyn-keyword/dyn-2021-edition-error.rs
+++ b/tests/ui/dyn-keyword/dyn-2021-edition-error.rs
@@ -7,6 +7,12 @@ fn function(x: &SomeTrait, y: Box<SomeTrait>) {
     //~^ ERROR expected a type, found a trait
 }
 
+// Regression test for <https://github.com/rust-lang/rust/issues/138211>.
+extern "C" {
+    fn foo() -> *const SomeTrait;
+    //~^ ERROR expected a type, found a trait
+}
+
 trait SomeTrait {}
 
 fn main() {}

--- a/tests/ui/dyn-keyword/dyn-2021-edition-error.stderr
+++ b/tests/ui/dyn-keyword/dyn-2021-edition-error.stderr
@@ -30,6 +30,17 @@ LL | fn function(x: &SomeTrait, y: Box<dyn SomeTrait>) {
    |                                   +++
 
 error[E0782]: expected a type, found a trait
+  --> $DIR/dyn-2021-edition-error.rs:12:24
+   |
+LL |     fn foo() -> *const SomeTrait;
+   |                        ^^^^^^^^^
+   |
+help: you can add the `dyn` keyword if you want a trait object
+   |
+LL |     fn foo() -> *const dyn SomeTrait;
+   |                        +++
+
+error[E0782]: expected a type, found a trait
   --> $DIR/dyn-2021-edition-error.rs:6:14
    |
 LL |     let _x: &SomeTrait = todo!();
@@ -40,6 +51,6 @@ help: you can add the `dyn` keyword if you want a trait object
 LL |     let _x: &dyn SomeTrait = todo!();
    |              +++
 
-error: aborting due to 3 previous errors
+error: aborting due to 4 previous errors
 
 For more information about this error, try `rustc --explain E0782`.

--- a/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.rs
+++ b/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.rs
@@ -31,4 +31,11 @@ where
     Trait:, {}
 //~^ ERROR expected a type, found a trait
 
+struct HasGenerics<T> {
+    f: Trait,
+    //~^ ERROR expected a type, found a trait
+    t: T,
+}
+
+
 fn main() {}

--- a/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.rs
+++ b/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.rs
@@ -13,7 +13,6 @@ struct Foo2 {
     //~^ ERROR expected a type, found a trait
 }
 
-
 enum Enum1 {
     A(Trait),
     //~^ ERROR expected a type, found a trait
@@ -26,5 +25,10 @@ enum Enum2 {
     //~^ ERROR expected a type, found a trait
 }
 
+// Regression test for <https://github.com/rust-lang/rust/issues/138229>.
+pub struct InWhereClause
+where
+    Trait:, {}
+//~^ ERROR expected a type, found a trait
 
 fn main() {}

--- a/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.stderr
+++ b/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.stderr
@@ -22,7 +22,7 @@ LL |     b: dyn Trait,
    |        +++
 
 error[E0782]: expected a type, found a trait
-  --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:18:7
+  --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:17:7
    |
 LL |     A(Trait),
    |       ^^^^^
@@ -34,7 +34,7 @@ LL ~     A(T),
    |
 
 error[E0782]: expected a type, found a trait
-  --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:25:7
+  --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:24:7
    |
 LL |     B(Trait),
    |       ^^^^^
@@ -46,6 +46,17 @@ LL |     A(u32),
 LL ~     B(T),
    |
 
-error: aborting due to 4 previous errors
+error[E0782]: expected a type, found a trait
+  --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:31:5
+   |
+LL |     Trait:, {}
+   |     ^^^^^
+   |
+help: you can add the `dyn` keyword if you want a trait object
+   |
+LL |     dyn Trait:, {}
+   |     +++
+
+error: aborting due to 5 previous errors
 
 For more information about this error, try `rustc --explain E0782`.

--- a/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.stderr
+++ b/tests/ui/dyn-keyword/suggest-struct-or-union-add-generic-impl-trait.stderr
@@ -47,6 +47,18 @@ LL ~     B(T),
    |
 
 error[E0782]: expected a type, found a trait
+  --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:35:8
+   |
+LL |     f: Trait,
+   |        ^^^^^
+   |
+help: you might be missing a type parameter
+   |
+LL ~ struct HasGenerics<T, U: Trait> {
+LL ~     f: U,
+   |
+
+error[E0782]: expected a type, found a trait
   --> $DIR/suggest-struct-or-union-add-generic-impl-trait.rs:31:5
    |
 LL |     Trait:, {}
@@ -57,6 +69,6 @@ help: you can add the `dyn` keyword if you want a trait object
 LL |     dyn Trait:, {}
    |     +++
 
-error: aborting due to 5 previous errors
+error: aborting due to 6 previous errors
 
 For more information about this error, try `rustc --explain E0782`.

